contracts/AelinUpFrontDeal.sol

```diff
@@ -0,0 +1,299 @@
+            }
+        }
+
+        _mint(msg.sender, purchaseTokenAmount);
+    }
+
+    function _purchasePoolTokensWithNft(
+        UpFrontPool memory _poolData,
+        IAelinPool.NftPurchaseList[] calldata _nftPurchaseList,
+        uint256 _purchaseTokenAmount
+    ) internal {
+        require(hasNftList, "pool does not have an NFT list");
+
+        uint256 maxPurchaseTokenAmount;
+
+        for (uint256 i = 0; i < _nftPurchaseList.length; i++) {
+            IAelinPool.NftPurchaseList memory nftPurchaseList = _nftPurchaseList[i];
+            address _collectionAddress = nftPurchaseList.collectionAddress;
+            uint256[] memory _tokenIds = nftPurchaseList.tokenIds;
+
+            IAelinPool.NftCollectionRules memory nftCollectionRules = nftCollectionDetails[_collectionAddress];
+            require(nftCollectionRules.collectionAddress == _collectionAddress, "collection not in the pool");
+
+            if (nftCollectionRules.purchaseAmountPerToken) {
+                maxPurchaseTokenAmount += nftCollectionRules.purchaseAmount * _tokenIds.length;
+            }
+
+            if (!nftCollectionRules.purchaseAmountPerToken && nftCollectionRules.purchaseAmount > 0) {
+                require(!nftWalletUsedForPurchase[_collectionAddress][msg.sender], "wallet already used for nft set");
+                nftWalletUsedForPurchase[_collectionAddress][msg.sender] = true;
+                maxPurchaseTokenAmount += nftCollectionRules.purchaseAmount;
+            }
+
+            if (nftCollectionRules.purchaseAmount == 0) {
+                maxPurchaseTokenAmount = _purchaseTokenAmount;
+            }
+
+            if (NftCheck.supports721(_collectionAddress)) {
+                _blackListCheck721(_collectionAddress, _tokenIds);
+            }
+            if (NftCheck.supports1155(_collectionAddress)) {
+                _eligibilityCheck1155(_collectionAddress, _tokenIds, nftCollectionRules);
+            }
+            if (_collectionAddress == CRYPTO_PUNKS) {
+                _blackListCheckPunks(_collectionAddress, _tokenIds);
+            }
+        }
+
+        require(_purchaseTokenAmount <= maxPurchaseTokenAmount, "purchase amount should be less the max allocation");
+
+        uint256 amountBefore = IERC20(_poolData.purchaseToken).balanceOf(address(this));
+        IERC20(_poolData.purchaseToken).transferFrom(msg.sender, address(this), _purchaseTokenAmount);
+        uint256 amountAfter = IERC20(_poolData.purchaseToken).balanceOf(address(this));
+        uint256 purchaseTokenAmount = amountAfter - amountBefore;
+
+        if (_poolData.purchaseTokenCap > 0) {
+            uint256 totalPoolAfter = totalSupply() + purchaseTokenAmount;
+            require(totalPoolAfter <= _poolData.purchaseTokenCap, "cap has been exceeded");
+            if (totalPoolAfter == _poolData.purchaseTokenCap) {
+                poolDepositComplete = true;
+            }
+        }
+
+        _mint(msg.sender, purchaseTokenAmount);
```

Comment:
Above comment applies here for pricing deal tokens and holding the value in a mapping until claiming the deal token